core/domain/exp_services.py

```diff
@@ -2063,6 +2063,248 @@ def get_interaction_id_for_state(exp_id, state_name):
+                user_services.get_checkpoints_in_order(
+                    latest_exploration.init_state_name,
+                    latest_exploration.states),
+                user_services.get_checkpoints_in_order(
+                    furthest_reached_exploration.init_state_name,
+                    furthest_reached_exploration.states),
+                model.furthest_reached_checkpoint_state_name
+            )
+    )
+
+    # If the most recently reached checkpoint doesn't exist in current
+    # exploration.
+    if (
+        most_recently_reached_checkpoint_in_current_exploration !=
+        model.most_recently_reached_checkpoint_state_name
+    ):
+        model.most_recently_reached_checkpoint_state_name = (
+            most_recently_reached_checkpoint_in_current_exploration)
+        model.most_recently_reached_checkpoint_exp_version = (
+            latest_exploration.version)
+        model.update_timestamps()
+        model.put()
+
+    # If the furthest reached checkpoint doesn't exist in current
+    # exploration.
+    if (
+        furthest_reached_checkpoint_in_current_exploration !=
+        model.furthest_reached_checkpoint_state_name
+    ):
+        model.furthest_reached_checkpoint_state_name = (
+            furthest_reached_checkpoint_in_current_exploration)
+        model.furthest_reached_checkpoint_exp_version = (
+            latest_exploration.version)
+        model.update_timestamps()
+        model.put()
+
+    return exp_fetchers.get_logged_out_user_progress(unique_progress_url_id)
+
+
+def sync_logged_out_learner_progress_with_logged_in_progress(
+    user_id, unique_progress_url_id):
+
+    """Syncs logged out and logged in learner progress."""
+
+    logged_out_user_data = (
+        exp_fetchers.get_logged_out_user_progress(unique_progress_url_id))
+
+    exp_id = logged_out_user_data.exploration_id
+    exp_data = exp_fetchers.get_exploration_by_id(exp_id)
+
+    logged_in_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exp_id)
+
+    if logged_in_user_model is None:
+        logged_in_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exp_id)
+
+        logged_in_user_model.most_recently_reached_checkpoint_exp_version = (
+            logged_out_user_data.most_recently_reached_checkpoint_exp_version
+        )
+        logged_in_user_model.most_recently_reached_checkpoint_state_name = (
+            logged_out_user_data.most_recently_reached_checkpoint_state_name
+        )
+        logged_in_user_model.furthest_reached_checkpoint_exp_version = (
+            logged_out_user_data.furthest_reached_checkpoint_exp_version
+        )
+        logged_in_user_model.furthest_reached_checkpoint_state_name = (
+            logged_out_user_data.furthest_reached_checkpoint_state_name
+        )
+        logged_in_user_model.update_timestamps()
+        logged_in_user_model.put()
+
+    else:
```

Comment:
I do not see it covering the scenerio of the exploration version of progress as logged in being different than that of the logged out one. We should definitely handle that scenerio